cache-filter/src/unique_callout.rs

```diff
@@ -0,0 +1,129 @@
+use crate::{info, warn};
+use proxy_wasm::{
+    hostcalls::{get_shared_data, set_shared_data},
+    types::Status,
+};
+use threescale::proxy::CacheKey;
+
+// Callout lock is acquired by placing a key-value pair inside shared data.
+// Since, only one thread is allowed to access shared data (host uses mutex) thus only one winner;
```

Comment:
Since this is far from trivial, this needs a TL;DR about the algorithm regardless of comments inside the function. Also remember that this is based on what the Envoy proxy is doing, since the spec does not go into details - so make sure to document these assumptions.

Reply:
Added a huge comment above explaining how setting and freeing is done for the lock.